pkg/cmd/server/startconfig/kube_connection_args.go

```diff
@@ -0,0 +1,117 @@
+package startconfig
+
+import (
+	"errors"
+	"net/url"
+
+	"github.com/spf13/pflag"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+
+	"github.com/openshift/origin/pkg/cmd/flagtypes"
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+)
+
+type KubeConnectionArgs struct {
+	KubernetesAddr flagtypes.Addr
+
+	// ClientConfig is used when connecting to Kubernetes from the master, or
+	// when connecting to the master from a detached node. If StartKube is true,
+	// this value is not used.
+	ClientConfig clientcmd.ClientConfig
+	// ClientConfigLoadingRules is the ruleset used to load the client config.
+	// Only the CommandLinePath is expected to be used.
+	ClientConfigLoadingRules clientcmd.ClientConfigLoadingRules
+
+	CertArgs *CertArgs
+}
+
+func BindKubeConnectionArgs(flags *pflag.FlagSet) *KubeConnectionArgs {
+	kubeConnectionCommandArgs := NewDefaultKubeConnectionArgs()
+
+	flags.Var(&kubeConnectionCommandArgs.KubernetesAddr, "kubernetes", "The address of the Kubernetes server (host, host:port, or URL). If specified, no Kubernetes components will be started.")
+	flags.StringVar(&kubeConnectionCommandArgs.ClientConfigLoadingRules.CommandLinePath, "kubeconfig", "", "Path to the kubeconfig file to use for requests to the Kubernetes API.")
+
+	return kubeConnectionCommandArgs
+}
+
+func NewDefaultKubeConnectionArgs() *KubeConnectionArgs {
+	config := &KubeConnectionArgs{}
+
+	config.KubernetesAddr = flagtypes.Addr{Value: "localhost:8443", DefaultScheme: "https", DefaultPort: 8443, AllowPrefix: true}.Default()
+	config.ClientConfig = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&config.ClientConfigLoadingRules, &clientcmd.ConfigOverrides{})
+
+	return config
+}
+
+func (cfg KubeConnectionArgs) GetExternalKubernetesClientConfig() (*client.Config, bool, error) {
+	if len(cfg.ClientConfigLoadingRules.CommandLinePath) == 0 || cfg.ClientConfig == nil {
+		return nil, false, nil
+	}
+	clientConfig, err := cfg.ClientConfig.ClientConfig()
+	if err != nil {
+		return nil, false, err
+	}
+	return clientConfig, true, nil
+}
+
+func (cfg KubeConnectionArgs) GetKubernetesAddress(defaultAddress *url.URL) (*url.URL, error) {
+	if cfg.KubernetesAddr.Provided {
```

Comment:
Does this mean we're back to `--kubernetes` trumping the hostname in `--kubeconfig`, and mean we might send credentials to the wrong server?